sound/soc/sof/topology.c

```diff
@@ -2214,28 +2224,52 @@ static int sof_link_hda_process(struct snd_sof_dev *sdev,
 		if (!dai->name)
 			continue;
 
-		if (strcmp(link->name, dai->name) == 0 &&
-		    dai->comp_dai.direction == direction) {
+		if (strcmp(link->name, dai->name) == 0) {
+			if (dai->comp_dai.direction ==
+			    SNDRV_PCM_STREAM_PLAYBACK) {
```

Comment:
@RanderWang hopefully this will be the last issue. Apart from checking the direction, you also need to check if tx_slot is non-zero to proceed here. Similarly for capture.

Reply:
@ranj063 I have thought about this. There is no need here
(1) if there is a playback dai then tx_slot should be correct, please see sof_link_hda_load.
     tx_slot would be queried if this dai link support playback. So I don't want to check it again.

(2) tx_slot could be zero and this means dma channel zero is used for hda.

Reply:
> @ranj063 I have thought about this. There is no need here
> (1) if there is a playback dai then tx_slot should be correct, please see sof_link_hda_load.
> tx_slot would be queried if this dai link support playback. So I don't want to check it again.
> 
> (2) tx_slot could be zero and this means dma channel zero is used for hda.

@RanderWang sorry what does it mean by "dma channel zero is used for hda"?
Also what about the capture case, is rx_slot is zero, you should be doing nothing but with this change you'd still send an ipc, isnt it?

Reply:
@RanderWang i think I get your point. Basically if rx_slot is non-zero, the expectation is there we will not be able to find dai with the same dai_link with a capture direction. Is that your argument?

Reply:
@ranj063  I just took playback for example. Capture is the same as playback.  tx_slot and rx_slot
could be zero. They could be only send if this dai link supports them

Reply:
This is the code I submitted a few days ago, but later I check it is redundant, so I removed it.
 
		if (strcmp(link->name, dai->name) == 0) {
			if (dai->comp_dai.direction ==
			    SNDRV_PCM_STREAM_PLAYBACK) {
				if (!link->dpcm_playback)
					return -EINVAL;

				config->hda.link_dma_ch = tx_slot;
				playback_found = 1;
			} else {
				if (!link->dpcm_capture)
					return -EINVAL;

				config->hda.link_dma_ch = rx_slot;
				capture_found = 1;
			}

Reply:
@RanderWang it makes sense. But on the other hand, I think we should be OK to have a BE link with playback and capture support but have no dai's that require it. So, I think we should remove the logic for the playback_foudn and capture_found too. Same argument for sof_set_dai_config() too.

Reply:
@ranj063  @keyonjie  I don't understand why. Ff no dai is found, how to restore config dai in FW ? For HDA, HDA dai could be set without dai index in dai.

Reply:
@RanderWang my point was that if dpcm_playback and dpcm_capture can be set for a BE dai link but that doesnt mean we should always be able to fine a dai for each one of those.

Reply:
For this case, there is a error in topology file. So we should report this error to user.

Reply:
@RanderWang i do not think this is an error. Basically, what this means is that the BE dai link supports both playback and capture but there is no need for it in the pipelines in topology.

Reply:
@ranj063  It is a good discussing. As I know dpcm_playback and dpcm_capture are set in machine driver, and they would be reported to ASOC. So If you cat /proc/asound/cardx, you could find corresponding nodes and also aplay -l. users could get this info to aplay audio. If there is no dai, error would be reported by aplay

Reply:
So if there is dai link without any dai supporting it, could do you share me the using case. I don't get it. Thanks.

Reply:
@RanderWang just take the yorp and remove the capture pipeline for SSP2. thats what I did.
Alternatively, I thnk we should also be able to remove playback and capture on SSP2 and still have just the dai_config for SSP2 BE dai link.

Reply:
@ranj063  But I still don't get why we would remove playback and capture on SSP2 and still have just the dai_config for SSP2 BE dai link.

Reply:
@RanderWang the point is you could have just playback or just capture on SSP2.

Reply:
you idea is:  playback or capture are defined in machine driver, but our topology could support it or not.  I discuss with Keyon, and we agree with you at this point. Our idea is: machine driver could define two dai, but topology should support at least one dai. So the error change in sof_set_dai_config is correct. It only check whether only one dai exist. And in hda function, only check one of playback or capture exsit. The code should be the same as sof_set_dai_config. What is your idea? thanks!

Reply:
@RanderWang @keyonjie I think there's no need to check even if one dai is set. The machine driver could define SSP2 but what if I do not want to use SSP2 at all for playback/capture?Logically, in that case i'd probably remove the dai_config for SSP2 but I could just leave it as well.

Reply:
@ranj063  if you do not want to use SSP2 at all , why this dai link is set in topology file? If it is not in topology file, these function could not be called.